<?xml version='1.0' encoding='UTF-8'?>
<root>
 <tabbed_widget parent="main_window" name="Main Window">
  <Tab tab_name="XY" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="XYPlot">
       <range right="379.612646" bottom="-43.542017" top="-26.059706" left="348.045749"/>
       <limitY/>
       <curve curve_x="/rko_lio/odometry/pose/pose/position/x" name="/rko_lio/odometry/pose/pose/position/[x;y]" curve_y="/rko_lio/odometry/pose/pose/position/y" color="#bcbd22"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="Z" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179163" bottom="-1.583488" top="-1.509110" left="12.339349"/>
       <limitY/>
       <curve name="/rko_lio/odometry/pose/pose/position/z" color="#1f77b4"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="RPY" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179163" bottom="-93.743798" top="3.408822" left="12.339349"/>
       <limitY/>
       <curve name="roll_deg" color="#d62728"/>
       <curve name="pitch_deg" color="#1ac938"/>
       <curve name="yaw_deg" color="#ff7f0e"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="ang.vel" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179163" bottom="-0.058730" top="0.048645" left="12.339349"/>
       <limitY/>
       <curve name="/rko_lio/odometry/twist/twist/angular/x" color="#f14cc1"/>
       <curve name="/rko_lio/odometry/twist/twist/angular/y" color="#9467bd"/>
       <curve name="/rko_lio/odometry/twist/twist/angular/z" color="#17becf"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="lin.vel" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179163" bottom="-0.754731" top="8.419569" left="12.339349"/>
       <limitY/>
       <curve name="/rko_lio/odometry/twist/twist/linear/x" color="#bcbd22"/>
       <curve name="/rko_lio/odometry/twist/twist/linear/y" color="#1f77b4"/>
       <curve name="/rko_lio/odometry/twist/twist/linear/z" color="#d62728"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="imu.accel" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179841" bottom="-1.548929" top="11.317272" left="12.207884"/>
       <limitY/>
       <curve name="/rko_lio/re_imu/linear_acceleration/x" color="#1ac938"/>
       <curve name="/rko_lio/re_imu/linear_acceleration/y" color="#ff7f0e"/>
       <curve name="/rko_lio/re_imu/linear_acceleration/z" color="#f14cc1"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <Tab tab_name="imu.ang.vel" containers="1">
   <Container>
    <DockSplitter orientation="-" count="1" sizes="1">
     <DockArea name="...">
      <plot style="Lines" flip_x="false" flip_y="false" mode="TimeSeries">
       <range right="17.179841" bottom="-0.064745" top="0.042774" left="12.207884"/>
       <limitY/>
       <curve name="/rko_lio/re_imu/angular_velocity/x" color="#9467bd"/>
       <curve name="/rko_lio/re_imu/angular_velocity/y" color="#17becf"/>
       <curve name="/rko_lio/re_imu/angular_velocity/z" color="#bcbd22"/>
      </plot>
     </DockArea>
    </DockSplitter>
   </Container>
  </Tab>
  <currentTabIndex index="6"/>
 </tabbed_widget>
 <use_relative_time_offset enabled="1"/>
 <!-- - - - - - - - - - - - - - - -->
 <!-- - - - - - - - - - - - - - - -->
 <Plugins>
  <plugin ID="DataLoad CSV">
   <parameters time_axis="" delimiter="0"/>
  </plugin>
  <plugin ID="DataLoad MCAP"/>
  <plugin ID="DataLoad ROS2 bags">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
   <selected_topics value=""/>
  </plugin>
  <plugin ID="DataLoad ULog"/>
  <plugin ID="ROS2 Topic Subscriber">
   <use_header_stamp value="false"/>
   <discard_large_arrays value="true"/>
   <max_array_size value="100"/>
   <boolean_strings_to_number value="true"/>
   <remove_suffix_from_strings value="true"/>
   <selected_topics value="/rko_lio/accel;/rko_lio/odometry;/rko_lio/re_imu;/rko_lio/ang_vel_line;/rko_lio/body_accel_line"/>
  </plugin>
  <plugin ID="UDP Server"/>
  <plugin ID="WebSocket Server"/>
  <plugin ID="ZMQ Subscriber"/>
  <plugin ID="Fast Fourier Transform"/>
  <plugin ID="Quaternion to RPY"/>
  <plugin ID="Reactive Script Editor">
   <library code="--[[ Helper function to create a series from arrays&#xa;&#xa; new_series: a series previously created with ScatterXY.new(name)&#xa; prefix:     prefix of the timeseries, before the index of the array&#xa; suffix_X:   suffix to complete the name of the series containing the X value. If [nil], use the index of the array.&#xa; suffix_Y:   suffix to complete the name of the series containing the Y value&#xa; timestamp:   usually the tracker_time variable&#xa;              &#xa; Example:&#xa; &#xa; Assuming we have multiple series in the form:&#xa; &#xa;   /trajectory/node.{X}/position/x&#xa;   /trajectory/node.{X}/position/y&#xa;   &#xa; where {N} is the index of the array (integer). We can create a reactive series from the array with:&#xa; &#xa;   new_series = ScatterXY.new(&quot;my_trajectory&quot;) &#xa;   CreateSeriesFromArray( new_series, &quot;/trajectory/node&quot;, &quot;position/x&quot;, &quot;position/y&quot;, tracker_time );&#xa;--]]&#xa;&#xa;function CreateSeriesFromArray( new_series, prefix, suffix_X, suffix_Y, timestamp )&#xa;  &#xa;  --- clear previous values&#xa;  new_series:clear()&#xa;  &#xa;  --- Append points to new_series&#xa;  index = 0&#xa;  while(true) do&#xa;&#xa;    x = index;&#xa;    -- if not nil, get the X coordinate from a series&#xa;    if suffix_X ~= nil then &#xa;      series_x = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_X) )&#xa;      if series_x == nil then break end&#xa;      x = series_x:atTime(timestamp)&#x9; &#xa;    end&#xa;    &#xa;    series_y = TimeseriesView.find( string.format( &quot;%s.%d/%s&quot;, prefix, index, suffix_Y) )&#xa;    if series_y == nil then break end &#xa;    y = series_y:atTime(timestamp)&#xa;    &#xa;    new_series:push_back(x,y)&#xa;    index = index+1&#xa;  end&#xa;end&#xa;&#xa;--[[ Similar to the built-in function GetSeriesNames(), but select only the names with a give prefix. --]]&#xa;&#xa;function GetSeriesNamesByPrefix(prefix)&#xa;  -- GetSeriesNames(9 is a built-in function&#xa;  all_names = GetSeriesNames()&#xa;  filtered_names = {}&#xa;  for i, name in ipairs(all_names)  do&#xa;    -- check the prefix&#xa;    if name:find(prefix, 1, #prefix) then&#xa;      table.insert(filtered_names, name);&#xa;    end&#xa;  end&#xa;  return filtered_names&#xa;end&#xa;&#xa;--[[ Modify an existing series, applying offsets to all their X and Y values&#xa;&#xa; series: an existing timeseries, obtained with TimeseriesView.find(name)&#xa; delta_x: offset to apply to each x value&#xa; delta_y: offset to apply to each y value &#xa;  &#xa;--]]&#xa;&#xa;function ApplyOffsetInPlace(series, delta_x, delta_y)&#xa;  -- use C++ indeces, not Lua indeces&#xa;  for index=0, series:size()-1 do&#xa;    x,y = series:at(index)&#xa;    series:set(index, x + delta_x, y + delta_y)&#xa;  end&#xa;end&#xa;"/>
   <scripts/>
  </plugin>
  <plugin ID="CSV Exporter"/>
  <plugin ID="ROS2 Topic Re-Publisher"/>
 </Plugins>
 <!-- - - - - - - - - - - - - - - -->
 <previouslyLoaded_Datafiles/>
 <previouslyLoaded_Streamer name="ROS2 Topic Subscriber"/>
 <!-- - - - - - - - - - - - - - - -->
 <customMathEquations>
  <snippet name="yaw_deg">
   <global></global>
   <function>return value * 180/3.14159</function>
   <linked_source>/rko_lio/odometry/pose/pose/orientation/yaw</linked_source>
  </snippet>
  <snippet name="pitch_deg">
   <global></global>
   <function>return value * 180/3.14159</function>
   <linked_source>/rko_lio/odometry/pose/pose/orientation/pitch</linked_source>
  </snippet>
  <snippet name="roll_deg">
   <global></global>
   <function>return value * 180/3.14159</function>
   <linked_source>/rko_lio/odometry/pose/pose/orientation/roll</linked_source>
  </snippet>
 </customMathEquations>
 <snippets/>
 <!-- - - - - - - - - - - - - - - -->
</root>

